### ISSUE TITLE
LOOT is unable to handle "Training costs" plugins

### ISSUE THREAD

**Issue by pStyl3** [loot/loot#971]
[Training costs](https://www.nexusmods.com/skyrim/mods/2746)

This Skyrim mod by the (banned) author "moonballz" contains the following plugins:
> training_costs_1_x1_x1.esp
training_costs_5_x1_x1.esp
training_costs_5_x2_x3.esp
training_costs_10_x1_x1.esp
training_costs_10_x2_x3.esp
training_costs_20_x1_x1.esp
training_costs_20_x2_x3.esp

These plugins were uploaded not even a month after Skyrim's release .. it's likely that these weren't made with xEdit (and certainly not with the Creation Kit), what would make TESVSnip the other possibility.
And if that's the case, I think we should warn the users of using these plugins (as plugins made with Snip can break the game).

In fact if I try to open them with xEdit, I get a lot of error messages.
> NodeDataSize not initialized. (...\TES5Edit\VirtualTrees.pas, line 29917)
Invalid pointer operation.
Access violation at address 0058DDD4 in module 'TES5Edit.exe'. Write of address 005488D8.

That being said, the issue for LOOT is that if the user tries to sort his/her load order with one or more of these plugins being present, LOOT errors out if the user wants to Apply or Cancel the new load order. The window with the message "Oh no, something went wrong! You can check your LOOTDebugLog.txt (you can get to it through the main menu) for more information." appears for example if Cancel is used, but LOOT does then not leave this state and must be closed and restarted.

LOOTDebugLog.txt
>[12:58:45.558985] [trace]: HighResTexturePack03.esp: Caching CRC value.
[12:58:45.559011] [error]: Cannot read plugin file "training_costs_5_x1_x1.esp". Details: C:\Games\Steam\steamapps\common\Skyrim\Data\training_costs_5_x1_x1.esp : esplugin error code: 5
[12:58:45.559036] [trace]: HighResTexturePack02.esp: Plugin loading complete.
[12:58:45.559074] [trace]: Loading Dawnguard.esm
[12:58:45.559108] [trace]: Calculating CRC for: C:\Games\Steam\steamapps\common\Skyrim\Data\HighResTexturePack03.esp
[12:58:45.559136] [error]: Caught exception while trying to add training_costs_5_x1_x1.esp to the cache: Cannot read "training_costs_5_x1_x1.esp". Details: C:\Games\Steam\steamapps\common\Skyrim\Data\training_costs_5_x1_x1.esp : esplugin error code: 5
[12:58:45.559169] [trace]: Loading Dragonborn.esm

and
> 12:58:47.404696] [error]: Exception while executing query: No plugin "training_costs_5_x1_x1.esp" exists.
[12:58:47.405466] [error]: Chromium console message from http://loot/ui/app.bundle.js at line 1: Error: Oh no, something went wrong! You can check your LOOTDebugLog.txt (you can get to it through the main menu) for more information.
    at onFailure (http://loot/ui/app.bundle.js:1:462892)

Or when Apply is used:
> [12:24:26.761238] [error]: libloadorder failed to set the load order. Details: The plugin "training_costs_20_x2_x3.esp" is not in the load order
[12:24:26.761384] [error]: Exception while executing query: libloadorder failed to set the load order. Details: The plugin "training_costs_20_x2_x3.esp" is not in the load order: Libloadorder error

**Comment by WrinklyNinja** [comment]
Interesting, if I install all the plugins and try to sort, it fails earlier, so the apply/cancel buttons don't show up. Can you upload your debug log for the behaviour you see?

**Comment by pStyl3** [comment]
[Part 1](https://pastebin.com/raw/ubRYe190) and [Part 2](https://pastebin.com/raw/MPjB3XFc)

That is after I sorted and clicked on the Cancel button.

**Comment by WrinklyNinja** [comment]
Huh, I can replicate it now, not sure what was going on...

OK, so applying is failing because LOOT parses the training costs plugins during sorting and sees they're corrupt, so it removes them from its plugin cache, then it tries to set the load order for the rest of the plugins. However, libloadorder doesn't know the training costs plugins are corrupt, because it only parses plugin headers, so it complains that the given load order is incomplete. Libloadorder probably needs to check more fully to avoid this.

As for cancelling, that's failing because LOOT gets the current load order (which includes the training costs plugins) then for each one tries to get its cached plugin object, but they were removed for the training costs plugins because they're corrupt, so an exception is thrown. Getting libloadorder to ignore the plugins and so not return them as part of the load order would avoid this, but LOOT should probably be catching that exception since it's reasonable that plugin objects might not exist.

**Comment by WrinklyNinja** [comment]
OK, [this build](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_0.13.1-1-g47a0275_sorting-fix.7z) should fix cancelling the load order.

**Comment by pStyl3** [comment]
I can confirm that cancelling the load order works with that build.

**Comment by pStyl3** [comment]
moonballz uploaded another mod sorta like this (upload date is one day before the upload of training costs): [Levelling pace](https://www.nexusmods.com/skyrim/mods/2639)

It contains the following plugins:
> pace_A_75_25 (vanilla Skyrim).esp
pace_B_116_24.esp
pace_C_198_22.esp
pace_D_280_20.esp
pace_E_485_15.esp
pace_F_690_10.esp
pace_G_895_5.esp
pace_H_1100_0 (linear).esp

In the comment section of the mod, moonballz writes:
> With TESsnip it's easy to change.

As such, he indeed used Snip for Training costs and Levelling pace, and we should add all these plugins to the masterlist with an message of `type: error`.

**Comment by WrinklyNinja** [comment]
I've done most of the necessary work for this, but there are a few outstanding issues.

As things are, when LOOT fully loads plugins and finds one is invalid, it'll remove it from its cache. This has knock-on effects because in various places it's assumed that this doesn't happen. I've fixed those assumptions, but I still need to think about whether the behaviour I went for is best. Also, fixing them means that the plugin effectively disappears from LOOT's UI and operations, which could be confusing. I can potentially add a general error message for any such plugins to let the user know why that happened.

Alternatively, I've made some performance improvements to plugin parsing, and fully parsing plugins is ~ 4.5x faster than before, so it might be fast enough now that plugins can be fully parsed at startup, removing all the confusion of thinking a plugin was valid then realising it isn't. I've got to do a bit more profiling for this though, and I'll probably put out a build with this change to get some feedback from people with slower hard drives.

**Comment by pStyl3** [comment]
Is there already a build I should test?

Also, please have a look at Nexus comments, the discussion with nova_25. There are one Skyrim and one Oblivion plugins which LOOT has problems with aswell (I will send you the Skyrim one, once I have it). I guess those whould be handled best in their own, seperate issues right?

**Comment by WrinklyNinja** [comment]
I think this should be fixed as of [this build](https://bintray.com/wrinklyninja/loot/download_file?file_path=loot_0.13.1-4-ga518103_dev.7z)? I think I just left it so that invalid plugins would disappear from the UI when plugins are fully loaded (sorting and conflict filtering).

**Comment by WrinklyNinja** [comment]
Also, `Quest Award Leveller - Knights of the Nine.esp` appears corrupt to esplugin:

```
[21:27:24.230212] [error]: Cannot read plugin file "Quest Award Leveller - Knights of the Nine.esp". Details: C:\Games\Steam\steamapps\common\Oblivion\Data\Quest Award Leveller - Knights of the Nine.esp : esplugin error code: 5
```

I've got no idea if this is correct or a bug though, it's too large a plugin for me to just spot it in a hex editor, I'll have to do some debugging. If it is correct, then it's causing the same problem as the training costs plugins, so it's probably not worth a separate issue.

**Comment by WrinklyNinja** [comment]
Yes, that should be possible, though I'm not sure when I'll have time to do it. I'll keep this issue open until it's done though.

I looked into the QAL - KOTN plugin and as your screenshot on the Nexus thread shows, it's got a few CELL records that aren't in a top-level group. esplugin expects all records to belong to a group, as that's what's documented as the format on the UESPWiki pages and I don't recall seeing top-level records anywhere before. I'm not sure if this counts as being corrupt or just very non-standard though, I guess I'll have to make some visible edits to those records and check if they appear in game. For the purposes of this issue it's corrupt though, and anything else is being tracked in WrinklyNinja/esplugin#8.

**Comment by WrinklyNinja** [comment]
I finally got around to testing the QAL - KOTN plugin out, and:

- Viewing the CELL records that aren't in a group in TES4Edit is broken, it displays data in the wrong fields.
- Editing the CELL records that aren't in a group in TES4Edit is also broken, I tried setting water height and it added a Climate subrecord instead.
- TES4Edit can't save edits to the out-of-group CELL records.
- Loading the plugin in the TES4 Construction Set shows that it detects that the plugin is editing those cells, but if I save the plugin it puts the cells into the group where they belong.

Given that the out-of-group CELL records don't seem to make any changes to the records, and that I can't reasonably make any changes to them while preserving their out-of-group-ness, I consider them to be corrupt wild edits.
